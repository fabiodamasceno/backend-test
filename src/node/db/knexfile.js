const baseSetting = {
  client: 'sqlite3',
  connection: {filename: ''},
  pool: 1,
  migrations: {
    tableName: 'migrations',
    directory: `${__dirname}/../db/migrations`
  },
  useNullAsDefault: true
}

const baseSettingsWith = filename => Object
  .assign({}, baseSetting, {connection: {filename}})

module.exports = {
  test: baseSettingsWith('./db.test.sqlite3'),
  development: baseSettingsWith('./db.development.sqlite3'),
  production: baseSettingsWith('./db.production.sqlite3'),
}
