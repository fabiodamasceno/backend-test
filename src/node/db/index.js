import knexBuild from 'knex'
import knexfile from  './knexfile'

import bookshelfBuild from 'bookshelf'
import validate from './validate'
import cascade from 'bookshelf-cascade-delete'

// istanbul ignore next: environment should be always test for istanbul
const env = process.env.NODE_ENV || 'development'

export const knex = knexBuild(knexfile[env])

const bookshelf = bookshelfBuild(knex)
bookshelf.plugin(validate)
bookshelf.plugin(cascade)

export default bookshelf