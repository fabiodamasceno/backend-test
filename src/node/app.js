import Koa from 'koa'
import bodyParser from 'koa-bodyparser'

import error_handler from './infrastructure/http/error_handler'

import venues from './venues'

function not_found(ctx) {
  if (ctx.body)
    return
  ctx.status = 404
  ctx.body = { error: 'Page not found' }
}

const app = new Koa()
  .use(error_handler)
  .use(bodyParser())
  .use(venues.routes())
  .use(not_found)

export default app