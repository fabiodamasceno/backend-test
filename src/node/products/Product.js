import db from '../db'
import Joi from 'joi'

import Item from '../items/Item'

const price = Joi.number().required()
const item_id = Joi.number().greater(0)

const schema = Joi.object().keys({price})

const Product = db.Model.extend({
  tableName: 'products',
  item: function() {
    return this.belongsTo(Item)
  },
  schema: {
    create: schema.keys({item_id: item_id.required()}),
    update: schema.keys({item_id})
  }
})

export default Product