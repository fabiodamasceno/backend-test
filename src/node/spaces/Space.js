import db from '../db'
import Joi from 'joi'

import Item from '../items/Item'

const hour_price = Joi.number().required()
const item_id = Joi.number().greater(0)

const schema = Joi.object().keys({ hour_price })

const Space = db.Model.extend({
  tableName: 'spaces',
  item: function() {
    return this.belongsTo(Item)
  },
  schema: {
    create: schema.keys({item_id: item_id.required()}),
    update: schema.keys({item_id})
  }
})

export default Space