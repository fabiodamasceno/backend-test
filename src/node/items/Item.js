import db from '../db'
import Joi from 'joi'

import Venue from '../venues/Venue'
import Space from '../spaces/Space'
import Product from '../products/Product'

const name = Joi.string().required()
const venue_id = Joi.number().greater(0)

const schema = Joi.object().keys({ name })

const Item = db.Model.extend({
  tableName: 'items',
  venue: function() {
    return this.belongsTo(Venue)
  },
  space: function() {
    return this.hasOne(Space)
  },
  product: function() {
    return this.hasOne(Product)
  },
  schema: {
    create: schema.keys({venue_id: venue_id.required()}),
    update: schema.keys({venue_id})
  }
}, {dependents: ['space', 'product']})

export default Item