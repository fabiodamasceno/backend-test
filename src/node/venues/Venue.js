import db from '../db'
import Joi from 'joi'

import Item from '../items/Item'

const name = Joi.string().required()
export const schema = Joi.object().keys({ name })

const Venue = db.Model.extend({
  tableName: 'venues',
  items: function() {
    return this.hasMany(Item)
  },
  schema: {
    create: schema,
    update: schema
  }
}, {dependents: ['items']})

export default Venue