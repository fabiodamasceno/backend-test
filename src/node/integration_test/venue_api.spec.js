import db, {knex} from '../db'
import Venue from '../venues/Venue'
import Item from '../items/Item'
import Product from '../products/Product'
import Space from '../spaces/Space'

import supertest from 'supertest'

import app from '../app'

async function createVenue() {
  let venue = await db.transaction(async transacting => {
    let v = await new Venue({name: 'venue'}).save(null, {transacting})

    let product_item = await new Item({name: 'product item'})
      .save({venue_id: v.id}, {transacting})

    let space_item = await new Item({name: 'space item'})
      .save({venue_id: v.id}, {transacting})

    await new Space({hour_price: 25.98})
      .save({'item_id': space_item.id}, {transacting})

    await new Product({price: 75.12})
      .save({'item_id': product_item.id}, {transacting})

    return v
  })
  return venue.toJSON()
}

describe('venues api',() => {

  const request = supertest.agent(app.listen())

  beforeEach(() => {
    return Promise.all([
      knex.raw('delete from `spaces`'),
      knex.raw('delete from `products`'),
      knex.raw('delete from `items`'),
      knex.raw('delete from `venues`')
    ])
  })

  it('GET /not_found', async () => {
    let res = await request.get('/not_found')
    expect(res.status).toBe(404)
    expect(res.body).toMatchSnapshot()
  })

  it('GET /venues', async () => {
    await createVenue()
    let res = await request.get('/venues')
    expect(res.status).toBe(200)
    expect(res.body).toMatchSnapshot()
  })

  it('GET /venues/:id', async () => {
    let venue = await createVenue()
    let res = await request.get(`/venues/${venue.id}`)
    expect(res.status).toBe(200)
    expect(res.body).toMatchSnapshot()
  })

  it('POST /venues', async () => {
    let res = await request.post('/venues').send({name: 'venue name'})
    expect(res.status).toBe(200)
    expect(res.body).toMatchSnapshot()
  })

  it('PATCH /venues/:id', async () => {
    let venue = await createVenue()
    let name = `${venue.name}_updated`
    let res = await request.patch(`/venues/${venue.id}`).send({name})
    expect(res.status).toBe(200)
    expect(res.body).toMatchSnapshot()
  })

  it('DELETE /venues/:id', async () => {
    let venue = await createVenue()
    let res = await request.delete(`/venues/${venue.id}`)
    expect(res.status).toBe(200)
    expect(res.body).toMatchSnapshot()
  })

})