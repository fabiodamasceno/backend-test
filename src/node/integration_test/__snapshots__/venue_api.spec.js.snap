// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`venues api DELETE /venues/:id 1`] = `
Object {
  "info": "success",
}
`;

exports[`venues api GET /not_found 1`] = `
Object {
  "error": "Page not found",
}
`;

exports[`venues api GET /venues 1`] = `
Array [
  Object {
    "id": 14672,
    "items": Array [
      "product item",
      "space item",
    ],
    "name": "venue",
  },
]
`;

exports[`venues api GET /venues/:id 1`] = `
Object {
  "id": 14673,
  "items": Array [
    Object {
      "id": 28654,
      "name": "product item",
      "product": Object {
        "id": 16730,
        "item_id": 28654,
        "price": 75.12,
      },
      "space": Object {},
      "venue_id": 14673,
    },
    Object {
      "id": 28655,
      "name": "space item",
      "product": Object {},
      "space": Object {
        "hour_price": 25.98,
        "id": 21917,
        "item_id": 28655,
      },
      "venue_id": 14673,
    },
  ],
  "name": "venue",
}
`;

exports[`venues api PATCH /venues/:id 1`] = `
Object {
  "id": 14675,
  "items": Array [
    Object {
      "id": 28656,
      "name": "product item",
      "product": Object {
        "id": 16731,
        "item_id": 28656,
        "price": 75.12,
      },
      "space": Object {},
      "venue_id": 14675,
    },
    Object {
      "id": 28657,
      "name": "space item",
      "product": Object {},
      "space": Object {
        "hour_price": 25.98,
        "id": 21918,
        "item_id": 28657,
      },
      "venue_id": 14675,
    },
  ],
  "name": "venue_updated",
}
`;

exports[`venues api POST /venues 1`] = `
Object {
  "id": 14674,
  "name": "venue name",
}
`;
